### DISCUSSION TITLE
BaseModel that acts as a dict

### DISCUSSION THREAD

**Discussion by kosciej16** [pydantic/pydantic#2948] [Question]
Is it possible to use some of pydantics magics to have that model contain just simple dict and act like that dict? e.g.

```
class Model(BaseModel):
    d: Dict[int, AnotherModel]

    def __getitem__(self, k):
        return self.d[k]
   
    def items(self):
        return self.d.items()

   # All other dict functionality
```
I tried mutliple inheritance but it does not work. I spend some time searching but couldn't find any helpful topic.

Thanks!

**Comment by PrettyWood** [comment]
Hi @kosciej16 
How about something like
```py
class Model(BaseModel):
    d: dict[int, Foo]

    def __getattr__(self, item):
        return getattr(self.d, item)
```
?

**Reply by kosciej16** [reply]
Thank you for a reply @PrettyWood 

I considered that, but it doesn't work for all dict methods (like __getitem__ or __delitem__), doesn't provide constructors (so additional code is needed) and breaks my IDE support.

**Comment by lesnik512** [comment]
I adopted my [solution](https://github.com/samuelcolvin/pydantic/discussions/2979#discussioncomment-1021949) for lists

But this solution do not validate assignments. It is interesting thing to play with, but I do not see practical cases of such pattern.

```python
from pydantic import BaseModel
from pydantic.generics import GenericModel
from typing import Dict, TypeVar, Generic

DataX = TypeVar('DataX')
DataY = TypeVar('DataY')

class DictModel(GenericModel, Generic[DataX, DataY]):
    __root__: Dict[DataX, DataT]

    def __getitem__(self, key):
        return self.__root__[key]

    def __setitem__(self, key, value):
        self.__root__[key] = value
 
    def __delitem__(self, key):
        del self.__root__[key]

DictModelStr = DictModel[str, str]

item = DictModelStr(__root__={"lol": "kek", "foo": "bar"})
print(item["lol"])  # kek
item["new"] = 1
print(item.__root__)  # {'lol': 'kek', 'foo': 'bar', 'new': 1}
del item["new"]
print(item.__root__)  # {'lol': 'kek', 'foo': 'bar'}
```

**Reply by kosciej16** [reply]
Thanks!
Indeed it's very enlightening part of code for me. But you are right, without assignment validation I see not practical use too.
